osf/management/commands/migrate_pagecounter_data.py

```diff
@@ -0,0 +1,120 @@
+import datetime
+import logging
+
+from django.core.management.base import BaseCommand
+from django.db import connection
+
+from framework import sentry
+
+logger = logging.getLogger(__name__)
+
+
+LIMIT_CLAUSE = ' LIMIT %s);'
+NO_LIMIT_CLAUSE = ');'
+
+REVERSE_SQL_BASE = '''
+UPDATE osf_pagecounter PC
+SET
+    resource_id = NULL,
+    file_id = NULL,
+    version = NULL
+WHERE PC.id in (
+    select PC.id from osf_pagecounter PC
+        left outer join osf_guid Guid on Guid._id = split_part(PC._id, ':', 2)
+        left outer join osf_basefilenode File on File._id = split_part(PC._id, ':', 3)
+    where
+          resource_id IS NOT NULL OR
+          file_id IS NOT NULL AND
+          Guid._id IS NOT NULL AND File._id IS NOT NULL
+'''
+REVERSE_SQL = '{} {}'.format(REVERSE_SQL_BASE, NO_LIMIT_CLAUSE)
+REVERSE_SQL_LIMITED = '{} {}'.format(REVERSE_SQL_BASE, LIMIT_CLAUSE)
+
+FORWARD_SQL_BASE = '''
+    UPDATE osf_pagecounter PC
+    SET
+        action = split_part(PC._id, ':', 1),
+        resource_id = Guid.id,
+        file_id = File.id,
+        version = NULLIF(split_part(PC._id, ':', 4), '')::int
+    FROM osf_guid Guid, osf_basefilenode File
+        WHERE
+              Guid._id = split_part(PC._id, ':', 2) AND
+              File._id = split_part(PC._id, ':', 3) AND
+              PC.id in (
+                  select PC.id from osf_pagecounter PC
+                      left outer join osf_guid Guid on Guid._id = split_part(PC._id, ':', 2)
+                      left outer join osf_basefilenode File on File._id = split_part(PC._id, ':', 3)
+                  where (PC.resource_id is NULL or PC.file_id IS NULL) AND
+                        Guid._id IS NOT NULL AND File._id IS NOT NULL
+'''
+FORWARD_SQL = '{} {}'.format(FORWARD_SQL_BASE, NO_LIMIT_CLAUSE)
+FORWARD_SQL_LIMITED = '{} {}'.format(FORWARD_SQL_BASE, LIMIT_CLAUSE)
+
+COUNT_SQL = '''
+select count(PC.id)
+    from osf_pagecounter as PC
+    left join osf_guid Guid on Guid._id = split_part(PC._id, ':', 2)
+    left join osf_basefilenode File on File._id = split_part(PC._id, ':', 3)
+where (PC.resource_id is NULL or PC.file_id IS NULL) AND
+      Guid._id IS NOT NULL AND File._id IS NOT NULL;
+'''
+
+
+class Command(BaseCommand):
+    help = '''Does the work of the pagecounter migration so that it can be done incrementally when convenient.
+    You will either need to set the page_size large enough to get all of the records, or you will need to run the
+    script multiple times until it tells you that it is done.'''
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '--dry_run',
+            type=bool,
+            default=False,
+            help='Run queries but do not write files',
+        )
+        parser.add_argument(
+            '--page_size',
+            type=int,
+            default=10000,
```

Comment:
Any idea about upper bounds on this page_size limit before we start locking too many things?  I don't have a feel for this myself.

Reply:
Not really. I think that will end up being a devops consideration which could be altered over time (or even have different sizes for different times of the day or whatever) if need be. For my other script, Matt had suggested 10k, and that's what the original version of this did, so it seems like a decent shorthand. That's about 15-25 seconds of work on my computer.

Reply:
Did you change your docker settings to see these speeds? just curious so I can speed this up myself! 

I saw this:
500 | 3 min
10000 | 3 min
100000 | 5 min
500000 | 20 min

Reply:
Yes, I have a bunch of memory and maybe a couple of extra processors for Docker. I think it's 14-15 GB of RAM and 4 processors, but if I don't confirm that by Monday, ping me again.

Reply:
Yup, 15gb ram, 6 cpus, 3gb swap